### PR TITLE
Include format setting in the attribute data of SQS messages

### LINKED ISSUE #73: format setting should be included in the attribute data of a message
so that we can decouple the sender and reader instead of needing for both to agree on a format a-priori

e.g. https://github.com/nervous-systems/fink-nottle/wiki/sqs.tagged which includes a `:fink-nottle/tag` attribute with value one of `#{:edn, :json, :transit}`

### PR DESCRIPTION
**Pull request by ghost**
This should close issue #73.

**Changes**:
- SQS messages get an additional attribute including the used SerDes format when being sent
- that attribute will be read and used upon receival to automatically deserialize the payload the correct way
- remove obsolete settings resulting from the above changes at various places
- updated tests and documentation

### CODE REVIEW

File: src/clj_sqs_extended/aws/sqs.clj [hunk 1]
```diff
@@ -11,11 +11,14 @@
            [com.amazonaws.services.sqs.model
             CreateQueueRequest
             DeleteMessageRequest
+            MessageAttributeValue
             PurgeQueueRequest
             ReceiveMessageRequest
             SendMessageRequest]))
 
 
+(def ^:private clj-sqs-ext-format-attribute "SerdesFormat")
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: I'd give it some identifiable info, e.g. `clj-sqs-extended/SerdesFormat`

File: src/clj_sqs_extended/aws/sqs.clj [hunk 2]
```diff
@@ -104,18 +107,24 @@
     message    - The actual message/data to be sent
 
   Options:
-    format     - The format (currently :json or :transit) to serialize outgoing
-                 messages with (optional, default: :transit)"
+    format     - The format (currently :json or :transit) to serialize the outgoing
+                 message with (optional, default: :transit)"
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: it's in the `Options` section already, no need to say that it's `optional` again

File: src/clj_sqs_extended/aws/sqs.clj [hunk 3]
```diff
@@ -104,18 +107,24 @@
...
      :or   {format :transit}}]
-   (->> (serdes/serialize message format)
-        (SendMessageRequest. queue-url)
-        (.sendMessage sqs-client)
-        (.getMessageId))))
+   (let [request (->> (serdes/serialize message format)
+                      (SendMessageRequest. queue-url))
+         message-attribute-format (doto (MessageAttributeValue.)
+                                        (.withDataType "String")
+                                        (.withStringValue (str (name format))))]
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: this is repeated in `send-fifo-message`, refactor it out

File: src/clj_sqs_extended/aws/sqs.clj [hunk 4]
```diff
@@ -155,29 +169,40 @@
   (->> (DeleteMessageRequest. queue-url (:receiptHandle message))
        (.deleteMessage sqs-client)))
 
+(defn- get-serdes-format-attribute
+  [message]
+  (when message
+    (let [attributes (.getMessageAttributes message)
+          format (get attributes clj-sqs-ext-format-attribute)]
+      (keyword (.getStringValue format)))))
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj:177]: use `->` more often

File: src/clj_sqs_extended/aws/sqs.clj [hunk 5]
```diff
@@ -155,29 +169,40 @@
...
+
+(defn- extract-relevant-keys
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: 👍 for refactoring this out. one improvement is to name it more descriptively, e.g. `extract-relevant-keys-from-message`

File: src/clj_sqs_extended/aws/sqs.clj [hunk 6]
```diff
@@ -155,29 +169,40 @@
...
+    {:keys [wait-time]
+     :or   {wait-time 0}}]
+   (let [request (doto (ReceiveMessageRequest. queue-url)
+                   (.setWaitTimeSeconds (int wait-time))
+                   ;; WATCHOUT: Without the next line our custom SerdesFormat attribute will not be received!
+                   (.setAttributeNames ["All"])
+                   ;; WATCHOUT: The next line is a design choice to read one message at a time from the queue
+                   (.setMaxNumberOfMessages (int 1)))
+         response (.receiveMessage sqs-client request)
+         sqs-message (first (.getMessages response))
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: is this ok? we're getting messages (plural) but discarding the rest and only using `first`? is there something like `.getOneMessage` or something?

File: src/clj_sqs_extended/aws/sqs.clj [hunk 7]
```diff
@@ -155,29 +169,40 @@
...
+         message (extract-relevant-keys sqs-message)]
```
Comment (Quantisan) [inline src/clj_sqs_extended/aws/sqs.clj]: this whole let block could be refactored out into its own fn. it's doing a lot.

File: test/clj_sqs_extended/sqs_test.clj [hunk 8]
```diff
@@ -32,8 +30,7 @@
                                        @fixtures/test-queue-url
                                        test-message {:format format}))
           (let [response (sqs/receive-message @fixtures/test-sqs-ext-client
-                                              @fixtures/test-queue-url
-                                              {:format format})]
```
Comment (Quantisan) [inline test/clj_sqs_extended/sqs_test.clj:36]: ❤️

**Review by Quantisan — CHANGES_REQUESTED** [review]

File: src/clj_sqs_extended/aws/sqs.clj (hunk 1 above)
Comment (ghost) [reply src/clj_sqs_extended/aws/sqs.clj]: That was pretty much what I tried first until I learned that there are restrictions in place:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-message-metadata.html

How about `clj-sqs-extended-SerdesFormat` or `clj-sqs-extended-serdes-format`?

File: src/clj_sqs_extended/aws/sqs.clj (hunk 6 above)
Comment (ghost) [reply src/clj_sqs_extended/aws/sqs.clj]: We're using `(.setMaxNumberOfMessages (int 1)))` so I would hope the SDK follows and only gives us one message here.

File: test/clj_sqs_extended/sqs_test.clj (hunk 8 above)
Comment (ghost) [reply test/clj_sqs_extended/sqs_test.clj:36]: :smile:

File: src/clj_sqs_extended/aws/sqs.clj (hunk 6 above)
Comment (Quantisan) [reply src/clj_sqs_extended/aws/sqs.clj]: oh I missed that. could you leave a comment here please?

File: src/clj_sqs_extended/aws/sqs.clj (hunk 1 above)
Comment (Quantisan) [reply src/clj_sqs_extended/aws/sqs.clj]: could you note the reference here too?

I'd say `clj-sqs-extended.serdes-format`

Comment (ghost) [reply src/clj_sqs_extended/aws/sqs.clj]: Roger that.

File: src/clj_sqs_extended/aws/sqs.clj (hunk 6 above)
Comment (ghost) [reply src/clj_sqs_extended/aws/sqs.clj]: Sorry, where is "here"? :confused:

Comment (Quantisan) [reply src/clj_sqs_extended/aws/sqs.clj]: on L200

Comment (ghost) [reply src/clj_sqs_extended/aws/sqs.clj]: Ah, sure.

**Review by Quantisan — APPROVED** [review]

**Pull request merged**